pkg/ipam/subnet.go

```diff
@@ -103,83 +106,98 @@ func NewSubnet(name, cidrStr string, excludeIps []string) (*Subnet, error) {
 			V4FreeIPList:     IPRangeList{&IPRange{Start: IP(v4FirstIP), End: IP(v4LastIP)}},
 			V4ReleasedIPList: IPRangeList{},
 			V4ReservedIPList: convertExcludeIps(v4ExcludeIps),
-			V4PodToIP:        map[string]IP{},
+			V4NicToIP:        map[string]IP{},
 			V4IPToPod:        map[IP]string{},
 			V6CIDR:           cidrs[1],
 			V6FreeIPList:     IPRangeList{&IPRange{Start: IP(v6FirstIP), End: IP(v6LastIP)}},
 			V6ReleasedIPList: IPRangeList{},
 			V6ReservedIPList: convertExcludeIps(v6ExcludeIps),
-			V6PodToIP:        map[string]IP{},
+			V6NicToIP:        map[string]IP{},
 			V6IPToPod:        map[IP]string{},
 			MacToPod:         map[string]string{},
-			PodToMac:         map[string]string{},
+			NicToMac:         map[string]string{},
+			PodToNicList:     map[string][]string{},
 		}
 		subnet.joinFreeWithReserve()
 	}
 	return &subnet, nil
 }
 
-func (subnet *Subnet) GetRandomMac(podName string) string {
-	if mac, ok := subnet.PodToMac[podName]; ok {
+func (subnet *Subnet) GetRandomMac(podName, nicName string) string {
+	if mac, ok := subnet.NicToMac[nicName]; ok {
 		return mac
 	}
 	for {
 		mac := util.GenerateMac()
 		if _, ok := subnet.MacToPod[mac]; !ok {
 			subnet.MacToPod[mac] = podName
-			subnet.PodToMac[podName] = mac
+			subnet.NicToMac[nicName] = mac
 			return mac
 		}
 	}
 }
 
-func (subnet *Subnet) GetStaticMac(podName, mac string, checkConflict bool) error {
+func (subnet *Subnet) GetStaticMac(podName, nicName, mac string, checkConflict bool) error {
 	if checkConflict {
 		if p, ok := subnet.MacToPod[mac]; ok && p != podName {
 			return ErrConflict
 		}
 	}
 	subnet.MacToPod[mac] = podName
-	subnet.PodToMac[podName] = mac
+	subnet.NicToMac[nicName] = mac
 	return nil
 }
 
-func (subnet *Subnet) GetRandomAddress(podName string, skippedAddrs []string) (IP, IP, string, error) {
+func (subnet *Subnet) pushPodNic(podName, nicName string) {
+	if subnet.V4NicToIP[nicName] != "" || subnet.V6NicToIP[nicName] != "" || subnet.NicToMac[nicName] != "" {
+		subnet.PodToNicList[podName] = util.UniqString(append(subnet.PodToNicList[podName], nicName))
+	}
+}
+
+func (subnet *Subnet) popPodNic(podName, nicName string) {
+	subnet.PodToNicList[podName] = util.RemoveString(subnet.PodToNicList[podName], nicName)
+}
+
+func (subnet *Subnet) GetRandomAddress(podName, nicName string, skippedAddrs []string) (IP, IP, string, error) {
 	subnet.mutex.Lock()
-	defer subnet.mutex.Unlock()
+	defer func() {
+		subnet.pushPodNic(podName, nicName)
```

Comment:
If error is not nil, is it necessary to still push the nic?

Reply:
```
func (subnet *Subnet) pushPodNic(podName, nicName string) {
	if subnet.V4NicToIP[nicName] != "" || subnet.V6NicToIP[nicName] != "" || subnet.NicToMac[nicName] != "" {
		subnet.PodToNicList[podName] = util.UniqString(append(subnet.PodToNicList[podName], nicName))
	}
}
```
This method will check whether the nic info needs to be recorded, so I think it's ok.